### PR TITLE
🏗 Transform aliased configured components

### PR DESCRIPTION
**Pull request by alanorozco**
Partial for #26572.

## Transformer

`transform-inline-configure-component` finds `configureComponent(MyConstructor, {foo: 'bar'})` calls and:

1. Inlines imported `MyConstructor` in current scope.

2. Replaces `*.*.STATIC_CONG_.foo` access with its value as defined in config object `{foo: 'bar'}`.

3. Replaces `configureComponent()` call with inlined `MyConstructor`'s id.

For runtime use of `configureComponent` like:

```js
import {MyConstructor} from './my-ctor';
const TAG = 'amp-foo';
const MyWrappedConstructor = configureComponent(
  MyConstructor,
  {TAG, foo: 'foo'}
);
```

Composed class gets aliased, hoisted and its configuration replaced inline:

```js
import {something} from './imported-by-constructor';

class MyConstructor {
  constructor() {
    // references to config object are replaced here, from:
    // `console.log(this.STATIC_CONFIG_.foo)`
    console.log(_foo);

    // ids in global scope are kept:
    const TAG = TAG;

    // unset properties are replaced with undefined, so this would
    // get minified
    const something = undefined || 'default value';

    // destructuring keeps only the default value when prop undefined:
    // `const {bar = 'default value'} = this.STATIC_CONFIG_`
    const bar = 'default value';
  }
}

const _foo = 'foo';

const TAG = 'amp-foo';
const MyWrappedConstructor = MyConstructor;
```

### CODE REVIEW

File: build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+          fsPath.join(currentDirname, importPath)
+        );
+        const importedDirname = fsPath.dirname(importedFilename);
+        const relImportPath = fsPath.relative(currentDirname, importedDirname);
+
+        const replacedMember = 'STATIC_CONFIG_';
+
+        const properties = propsObj.get('properties');
+
+        const propValues = {};
```
Comment (erwinmombay) [inline build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: probly want prototypeless object here

File: build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+ */
+const fs = require('fs-extra');
+const fsPath = require('path');
+const tempy = require('tempy');
+const {addNamed} = require('@babel/helper-module-imports');
+const {transformFileSync, transformSync} = require('@babel/core');
+
+/**
+ * @fileoverview
+ * Finds `configureComponent(Ctor, {foo: 'bar'})` calls and:
```
Comment (kevinkimball) [inline build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: ctor = constructor? Took me a minute to figure out might be easier to write out

File: build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+          const id = program.scope.generateUidIdentifier(name);
+          program.scope.push({id, init: value, kind: 'const'});
+          propValues[name] = id;
+        }
+
+        const currentDirname = dirname(file.opts.filename);
+        const importedModule = join(currentDirname, importPath);
+
+        // TODO(alanorozco): sourcemaps
```
Comment (erwinmombay) [inline build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: Can you add a tracking issue for this, i believe you're possibly going to need resorcery and terser for this

File: build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    code: false,
+    ast: true,
+    sourceType: 'module',
+    plugins: [[transformRedefineInline, opts]],
+  });
+
+/**
+ * Replaces `configureComponent()` wrapping calls.
+ *
+ * @return {Object}
```
Comment (erwinmombay) [inline build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: `!Object`

File: build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+ * 3. Replaces `configureComponent(...)` call with identifier for inlined, static
+ *    `MyConstructor`.
+ */
+
+const calleeName = 'configureComponent';
+const replacedMember = 'STATIC_CONFIG_';
+
+/**
+ * Sub-plugin that transforms inlined file that exports wrapped constructor.
+ * @return {Object}
```
Comment (erwinmombay) [inline build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: `!Object`

File: build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js (hunk 3 above)
Comment (alanorozco) [reply build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: Added issue. Why `terser`? What am I missing :)

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-inline-configure-component/index.js]: https://github.com/ampproject/amphtml/pull/26779/files#diff-321517b76b9f18ac62dff6aa46ec79b1R102-R135

https://github.com/ampproject/amphtml/blob/master/build-system/compile/single-pass.js#L771-L803

i'm not sure but @kristoferbaxter and @jridgewell used it in these to fix up the source maps in conjunction with resorcery

**Pull request merged**